skrub/_interpolation_join.py

```diff
@@ -0,0 +1,328 @@
+        ignored.
+
+        Parameters
+        ----------
+        data : array-like
+            Ignored.
+        targets : array-like
+            Ignored.
+
+        Returns
+        -------
+        self : InterpolationJoin
+            Returns self.
+        """
+        del data, targets
+        self._check_inputs()
+        X_values = self.vectorizer_.fit_transform(
+            self.right_table.loc[:, self._right_on]
+        )
+        estimators = self._get_estimator_assignments()
+        self.estimators_ = joblib.Parallel(self.n_jobs, verbose=3)(
+            joblib.delayed(_fit)(
+                X_values,
+                self.right_table,
+                assignment["columns"],
+                assignment["estimator"],
+            )
+            for assignment in estimators
+        )
+        return self
+
+    def _check_inputs(self):
+        if self.vectorizer is None:
+            self.vectorizer_ = compose.ColumnTransformer([], remainder="passthrough")
+        else:
+            self.vectorizer_ = base.clone(self.vectorizer)
+        self._check_condition()
+
+    def _check_condition(self):
+        if self.on is not None:
+            if self.right_on is not None or self.left_on is not None:
+                raise ValueError("Please provide EITHER on, OR (left_on AND right_on)")
+            left_on, right_on = self.on, self.on
+        else:
+            if self.right_on is None or self.left_on is None:
+                raise ValueError("Please provide EITHER on, OR (left_on AND right_on)")
+            left_on, right_on = self.left_on, self.right_on
+        self._left_on = [left_on] if isinstance(left_on, str) else list(left_on)
+        self._right_on = [right_on] if isinstance(right_on, str) else list(right_on)
+
+    def transform(self, left_table):
+        """Transform a table by joining inferred values to it.
+
+        the values of the `left_on` columns in `left_table` are used to predict
+        likely values for the contents of a matching row in `self.right_table`.
+
+        Parameters
+        ----------
+        left_table : DataFrame
+            The table to transform.
+
+        Returns
+        -------
+        join : DataFrame
+            The result of the join between `left_table` and inferred rows from
+            ``self.right_table``.
+        """
+        X_values = self.vectorizer_.transform(left_table.loc[:, self._left_on])
+        interpolated_parts = joblib.Parallel(self.n_jobs, verbose=3)(
+            joblib.delayed(_predict)(
+                X_values, assignment["columns"], assignment["estimator"]
+            )
+            for assignment in self.estimators_
+        )
+        interpolated_parts = _add_column_name_suffix(interpolated_parts, self.suffix)
+        original_index = left_table.index
+        return pd.concat(
+            [left_table.reset_index(drop=True)] + interpolated_parts, axis=1
+        ).set_index(original_index)
```

Comment:
that seems like we're reimplementing pd.concat? also it's not strictly the same because when we append many columns 1 by 1 we end up with a very fragmented dataframe:
```python
import pandas as pd

df1 = pd.DataFrame({c: range(3) for c in "ABC"})
df2 = pd.DataFrame({c: range(3) for c in "DEF"})
df3 = pd.concat([df1, df2])
print(df3._mgr.nblocks) # prints 2

df4 = df1.copy()
for col in df2:
    df4[col] = df2[col].values
print(df4._mgr.nblocks) # prints 4
```

(also it will consolidate the blocks if we reach 100 columns, though that probably wouldn't happen a lot)

that probably wouldn't be a problem though. still why do you think it's less readable to use concat and then restore the index?

Reply:
Nicely done on the fragmented analysis, I didn't have that in mind. I think that, in this situation, changing the index is error-prone and harder to interpret.
Could we at least agree on having `[left_table.reset_index(drop=True)] + interpolated_parts` on a separate line, for readability and debuggability?

Reply:
sure, or do you think it is easier to set the index in the dataframes we concatenate before concatenating? it avoids the reset_index and having to store the original index: https://github.com/skrub-data/skrub/pull/742/commits/31cf759c82e7ce0b8f7d148c3534f88c56d5a9ab